web01: Add lemmy

adisbladis opened the pull request:
~This is still WIP as the `pict-rs` (the Lemmy image storage server) setup isn't complete yet.~
~It technically works but we don't want to run it's embedded database on a samba file system, but instead store it locally and occasionally back up the database.~
_We're on a pict-rs 0.4 release candidate now that works for us._

The instance is up and running but set to private & defederated while we work out the kinks.

Closes https://github.com/nix-community/infra/issues/645.

Mic92 commented on hosts/web01/pgbackrest.nix:
```diff
@@ -0,0 +1,76 @@
+      global = {
+        repo1-retention-full = 3;
+        repo1-type = "cifs";
+        repo1-path = "/mnt/pgbackrest";
```
We also have a hetzner storage box for each physical machine, in case you need this (comes with borgbackup/cifs/sftp support).

Mic92 commented on hosts/web01/samba.nix:
```diff
@@ -0,0 +1,51 @@
+      };
+
+      "/mnt/pgbackrest" = {
+        device = "//${hostname}/u348918-sub2";
```
Ah. This seems to be the storage box.

zowoq commented on flake.nix:
```diff
@@ -42,6 +42,13 @@
+    # 0.4 comes with a separation of database and file storage which we require but isn't
+    # in a stable release yet.
+    # After pict-rs 0.4 is in nixpkgs we can remove this input.
+    pict-rs.url = "git+https://git.asonix.dog/asonix/pict-rs.git";
```
Is this safe to be updated twice a week or should it be pinned to a known good commit?

adisbladis replied:
I've pinned it to `0.4.0-rc11`. While I think it's _probably_ safe I don't want to stick any other admin with debugging this.

zowoq commented on hosts/web01/configuration.nix:
```diff
@@ -6,10 +6,29 @@
   networking.hostName = "web01";
   networking.hostId = "1cfd5aa3";
 
+  # Remove once https://github.com/NixOS/nixpkgs/pull/241651 can be used.
```
`xfs` was merged in https://github.com/NixOS/nixpkgs/pull/241526.

adisbladis replied:
Great. I had also made a PR but had missed that one.

zowoq replied:
Noticed that your PR has xfs enabled by default and the PR that was merged doesn't, need to enable it in srvos as well. https://github.com/numtide/srvos/pull/183

zowoq replied:
Updated our srvos input so this commit can be dropped. https://github.com/nix-community/infra/pull/660

adisbladis replied:
Dropped and rebased.

zowoq commented on hosts/web01/postgresql.nix:
```diff
@@ -0,0 +1,68 @@
+      # Advanced features
+      enable_partitionwise_join = "on";
+      enable_partitionwise_aggregate = "on";
+      jit = "on";
```
Doesn't `jit` require using `pkgs.postgresql_15_jit` instead of `pkgs.postgresql_15`?

adisbladis replied:
Indeed it does.

zowoq commented:
I have more questions/changes but probably easier to defer that for a bit while it's stabilized. As it's on a separate host so I'm fine with merging this as is.

adisbladis commented:
I'm running into bugs in `lemmy-ui` in how it's treating an internal image buffer (`Input buffer contains unsupported image format`).

I've added a [release candidate that fixes this issue to nixpkgs](https://github.com/NixOS/nixpkgs/pull/241858).
While we await that one getting merged I'm drafting this PR.

zimbatm commented on hosts/web01/lemmy.nix:
```diff
@@ -0,0 +1,47 @@
+  # Lemmy image storage
+  services.pict-rs = {
+    enable = true;
+    dataDir = "/mnt/lemmy-pict-rs";
```
It might be necessary to also extend the systemd unit to guarantee that the mount happens before the service gets started.

Mic92 replied:
`unitConfig.RequiresMountsFor = [ "" ];`

adisbladis replied:
I don't think it's required here as the mount point is systemd managed with autofs. I've added it anyway for good measure but I don't think it's strictly required.

adisbladis replied:
https://github.com/NixOS/nixpkgs/pull/243742

zowoq commented on modules/nixos/common.nix:
```diff
@@ -24,5 +24,7 @@
   # Just disable it since we are using telegraf to monitor raid health.
   systemd.services.mdmonitor.enable = false;
 
+  environment.systemPackages = [ pkgs.git ];
```
Added `gitMinimal` in https://github.com/numtide/srvos/commit/9f9073c808598b2584d509bfaab6a58e20d95ce1 so this can be dropped.

adisbladis commented:
I think this is ready to be merged.

zowoq commented on flake.nix:
```diff
@@ -42,6 +42,13 @@
+    # Pict-rs is the image hosting server for Lemmy.
+    # 0.4 comes with a separation of database and file storage which we require but isn't
+    # in a stable release yet.
+    # After pict-rs 0.4 is in nixpkgs we can remove this input.
```
When is this going to be done?

Mic92 replied:
Untested: https://github.com/NixOS/nixpkgs/pull/243218

adisbladis replied:
I also had a WIP branch for the same thing but I didn't know how to tackle the schema migration from 0.3 to 0.4 in the NixOS module.

adisbladis replied:
I've done a more comprehensive job in https://github.com/NixOS/nixpkgs/pull/243373.

adisbladis commented:
Marking as draft until https://github.com/NixOS/nixpkgs/pull/243373 & https://github.com/NixOS/nixpkgs/pull/243577 has hit unstable-small.

zowoq commented:
> Marking as draft until [NixOS/nixpkgs#243373](https://github.com/NixOS/nixpkgs/pull/243373) & [NixOS/nixpkgs#243577](https://github.com/NixOS/nixpkgs/pull/243577) has hit unstable-small.

With the staging-next merge we've now got breakage on darwin and aarch64-linux. https://github.com/nix-community/infra/pull/671 has the last unstable-small before the merge, it has pict-rs 0.4.0 and I've renamed swraid in srvos so we don't need the mkRenamedOptionModule fix.

adisbladis commented:
> With the staging-next merge we've now got breakage on darwin and aarch64-linux. https://github.com/nix-community/infra/pull/671 has the last unstable-small before the merge, it has pict-rs 0.4.0 and I've renamed swraid in srvos so we don't need the mkRenamedOptionModule fix.

Ok great. I've dropped the nixpkgs update from this PR and I'll rebase on #671 once that's been merged.

adisbladis commented:
@zowoq This has been rebased.

zowoq commented:
Feel free to merge.
